### ISSUE TITLE
tests: try harder to use locally built libs when running unit tests

### ISSUE THREAD

**Issue by tp-m** [mesonbuild/meson#1383]
There is a narrow scope to this issue, and a wider scope.

The problem is that if `LD_LIBRARY_PATH` is set, the tests run via `ninja test` might in some cases use libs from the `LD_LIBRARY_PATH` location rather than the locally-built libs, which might cause test failures.

Whether this happens or not seems to depend on the system/distro. Happens for me on debian sid, but apparently not for @nirbheek. In my case, the test binary has `RUNPATH` set (and is linked with `'-Wl,-rpath,...`. In @nirbheek's case `RPATH` is set.

It is not unusual that people still have and old version of libs installed in some prefix and they build the new version, test it, and only then install, so this is not an outlandish corner case IMO.

The narrow scope of this bug report is that `ninja test` fails in such circumstances where it picks up an old/wrong version of the lib via `LD_LIBRARY_PATH`. This could be "fixed" by meson prepending the bits it adds to the rpath to LD_LIBRARY_PATH when running tests.

The wider scope of this bug is that the problem exists for all exes built, and maybe meson can generally try harder to make the linker do the right thing so that the builddir is prefered over LD_LIBRARY_PATH (not sure that's always possible, I also haven't tested RPATH vs. RUNPATH on my system yet).

**Comment by nirbheek** [comment]
This needs more investigation, including a test case.

**Comment by nirbheek** [comment]
@tp-m can you try https://github.com/mesonbuild/meson/pull/1519 and see what you get? The test passes for me because `RPATH` overrides `LD_LIBRARY_PATH` on my system.

**Comment by tp-m** [comment]
@nirbheek The test fails for me with `ret was 2 instead of 1` - that's as intended/expected I presume?

**Comment by nirbheek** [comment]
Yep, that proves in a small test case that setting RUNPATH is unreliable.

**Comment by nirbheek** [comment]
Turns out we can force the linker to use `DT_RPATH` instead of `DT_RUNPATH` by passing `-Wl,--disable-new-dtags`. We should use this for test executables that aren't installed.

**Comment by jpakkane** [comment]
Don't we need to use it for every target that links to a library that we build ourselves but which also exists on the system?

**Comment by nirbheek** [comment]
Hmm, true, and we can't really change `DT_RPATH` to `DT_RUNPATH` by manually hacking the shared objects, nor can we know when to do that.

I guess we should indeed set `LD_LIBRARY_PATH` while running tests.

**Comment by jpakkane** [comment]
But again, no, all binaries need to be runnable from the build dir without envvar hacking. Anything else is terrible usability.

**Comment by nirbheek** [comment]
That already works, but obviously people are having issues with `LD_LIBRARY_PATH` being already set. The only way to override that is to prepend our own paths to it.

**Comment by nirbheek** [comment]
We could make that conditional on `LD_LIBRARY_PATH` being already set, I think that's a good middle-ground.

**Comment by bonzini** [comment]
Fixed by #11119.
